docs/fuel-sync-design.md

```diff
@@ -0,0 +1,345 @@
+proper handling of this case.
+
+#### Node behaviour
+
+All nodes have the same FS, so it is enough to describe the work of FS.
+
+##### Gossip about new data
+
+Each node gossips only information that is validated by itself and can be 
+proved by itself. It means the node gossips the block header only after 
+committing the block.
+
+The final destination of all blocks is BI. FS subscribes to events from 
+the BI service and gossips the information about the latest block to 
+the network on each commit via the `Gossiper` trait implemented by P2P.
+
+It gossips the blocks received from the network, from the block producer, and from the consensus.
+
+##### Ask about new data
+
+The network gossips a new block by default with an `ABGT` interval with 
+the [rule above](#gossip-about-new-data). But it is possible to have cases where 
+someone forgot to notify another node.
+
+FS has an internal timer that pings the P2P every `ABGT + Const`
+(for example, `Const = 5` seconds). This timer resets on each block 
+commit or timer tick. P2P, on each ping, asks neighbors about the latest block header.
+
+### The blocks synchronization on block header event
+
+FS has two synchronization phases, one is active, and another is pending. 
+Each phase has its range of blocks to sync. They don't overlap and are 
+linked. The purpose of the pending phase is to collect and join new 
+block headers while the active phase syncs the previous range. When 
+the first phase finishes synchronization, it takes some range from the 
+pending phase, joins with the remaining range from the active phase, 
+and requests blocks for the final range.
+
+Because each block header is sealed, and FS knows the block producer 
+of each block, it allows us to verify the validity of the header 
+and be sure that the block range is actual without fake heights.
+
+#### Fork
+
+If FS gets the valid sealed block header but already have another block for this height, it is a case of the fork.
+- For the PoA version: We ignore it and try to find blocks that link with our version of the blockchain. We should report it to the node manager(via logs, or maybe we want to do something more).
+- For the PoS version: We should have rules in the consensus for this case.
+
+#### Queuing of the block header
+
+When FS receives a block header, first FS checks:
+- Is it a valid header? It should be signed by the expected, for the node, producer/consensus at according height.
+  - No -> `Punisher::report`.
+  - Yes -> Next step.
+  - Don't know. The Ethereum part is not synced yet to this height. If the height is 
+  not far from the local ethereum height(the height returned by the 
+  ethereum node), we ignore this header and do not punish the sender.
+- Is it a new height?
+  - No -> Is it already in the blockchain?
+    - Yes -> Ignore(Maybe report some neutral error to P2P to prevent DDoS)
+    - No -> It is case of the [fork](#fork).
```

Comment:
hmm this one is a hard question. I'm not sure we can passively receive forks without being the case of a punishable action. Lets open a thread to discuss this one?

Reply:
I guess you are talking about "Yes -> Ignore(Maybe report some neutral error to P2P to prevent DDoS)".

Yes, we need to punish. Actually, it should be solved on the P2P level. Because P2P should punish nodes that send not requested information.

Each node can report about new height one time, each next time the reputation should decrease(if we didn't request it).